### DISCUSSION TITLE
the PathFromPoints class not find

### DISCUSSION THREAD

**Discussion by jengelbertt** [ManimCommunity/manim#3848] [Q&A] [answered]
This code is placed here by **MrDiver**. But it doesn't work for me. I do not have the specified classes(PathFromPoints,AttachHandlesAttachHandles). I use manim version 18. What should I do to solve the problem?

class TracePathExample(Scene):
    def construct(self):
        title = Tex("TracePathExample").to_edge(UP)
        self.add(title)

        xs = np.linspace(0,PI*2,5)
        def f(t):
            return [t,np.cos(t),0]
            # return np.random.random_integers(-2,2,size=3)
        ys = [f(t) for t in xs]
      
        path = PathFromPoints(ys).make_smooth()
        handles = AttachHandles(path)
        offset = ValueTracker(0)
        dots = []
        traces = []

        n = 2
        for i in range(n):
            def get(i):
                dot = Dot().add_updater(lambda m: m.move_to(path.point_from_proportion((offset.get_value()+i/n)%1)))
                trace = TracedPath(lambda : path.point_from_proportion((offset.get_value()+i/n)%1), dissipating_time=0)
                return dot,trace
            dot,trace = get(i)
            dots.append(dot)
            traces.append(trace)
        # square.add_updater(lambda m: m.become(square.animate.trace_path()))
        # self.add(path)
        self.add(dots[0], traces[0], handles)
        self.play(offset.animate.set_value(1), run_time=3, rate_func=linear)
        self.wait()

**Comment by uwezi** [comment]
a) I showed you yesterday how to correctly place code here....

![image](https://github.com/ManimCommunity/manim/assets/8582807/099ec21d-56dc-49de-8298-cde35fb746f3)


b) Where exactly, when and in which context did MrDiver post this code? (link?)

c) Here is a working version:
```py
class TracePathExample(Scene):
    def construct(self):
        title = Tex("TracePathExample").to_edge(UP)
        self.add(title)

        xs = np.linspace(0,PI*2,5)
        def f(t):
            return [t,np.cos(t),0]
            # return np.random.random_integers(-2,2,size=3)
        ys = [f(t) for t in xs]

        #path = PathFromPoints(ys).make_smooth()
        #handles = AttachHandles(path)
        path = VMobject().set_points_smoothly(ys)
        offset = ValueTracker(0)
        dots = []
        traces = []

        n = 2
        for i in range(n):
            def get(i):
                dot = Dot().add_updater(lambda m: m.move_to(path.point_from_proportion((offset.get_value()+i/n)%1)))
                trace = TracedPath(lambda : path.point_from_proportion((offset.get_value()+i/n)%1), dissipating_time=0)
                return dot,trace
            dot,trace = get(i)
            dots.append(dot)
            traces.append(trace)
        # square.add_updater(lambda m: m.become(square.animate.trace_path()))
        # self.add(path)
        self.add(dots[0], traces[0]) #, handles)
        self.play(offset.animate.set_value(1), run_time=3, rate_func=linear)
        self.wait()
```

https://github.com/ManimCommunity/manim/assets/8582807/0d625fd3-4deb-4c6f-991c-e98558b07130

**Reply by jengelbertt** [reply]
https://github.com/ManimCommunity/manim/discussions/3012

last example

**Reply by jengelbertt** [reply]
Thanks for correcting the code, but part of the diagram is not the same as the example.

**Reply by uwezi** [reply]
> Thanks for correcting the code, but part of the diagram is not the same as the example.

no, it is not the same, because I did not bother to also extract the handles of the individual Bezier curves. 

Also I didn't know what the example was, since those functions are not part of Manim and I had to guess from their names what they would possibly do...

The question is what you want to do with this example in the end, because it is only meant to highlight how the Bezier curves in a VMobject look like - and obviously MrDiver did forget to add the helper functions in his example, because they are not part of the Manim code base and they are hardly needed for most applications of Manim. The post is 2 years old, the codes have not yet made it into the documentation and might have been abandoned.

**Reply by uwezi** [reply]
> Thanks for correcting the code, but part of the diagram is not the same as the example.

this is a bit more like the "original" code - can we please continue on Discord? Github is not a great place for discussions...

```py
class TracePathExample(Scene):
    def construct(self):
        title = Tex("TracePathExample").to_edge(UP)
        self.add(title)

        xs = np.linspace(0,PI*2,5)
        def f(t):
            return [t,np.cos(t),0]
            # return np.random.random_integers(-2,2,size=3)
        ys = [f(t) for t in xs]

        #path = PathFromPoints(ys).make_smooth()
        path = VMobject().set_points_smoothly(ys)
        ps = path.get_anchors_and_handles()
        handles = VGroup(
            *[
                VGroup(
                    Dot(point=ps[0][i], color=RED),
                    Line(ps[0][i],ps[1][i],color=BLUE),
                    Line(ps[2][i],ps[3][i],color=GREEN),
                    Dot(point=ps[3][i], color=YELLOW),
                ) for i in range(len(ps[0]))]
        )
        offset = ValueTracker(0)
        dots = []
        traces = []

        n = 2
        for i in range(n):
            def get(i):
                dot = Dot().add_updater(lambda m: m.move_to(path.point_from_proportion((offset.get_value()+i/n)%1)))
                trace = TracedPath(lambda : path.point_from_proportion((offset.get_value()+i/n)%1), dissipating_time=0)
                return dot,trace
            dot,trace = get(i)
            dots.append(dot)
            traces.append(trace)
        # square.add_updater(lambda m: m.become(square.animate.trace_path()))
        # self.add(path)
        self.add(dots[0], traces[0], handles)
        self.play(offset.animate.set_value(1), run_time=3, rate_func=linear)
        self.wait()
  ```

https://github.com/ManimCommunity/manim/assets/8582807/c2dcd2c2-1cfc-4e3b-81c5-371ab46eac7c

**Reply by jengelbertt** [reply, marked as answer]
I am glad that manim community has a person like you. Thank you. you are amazing

**Reply by uwezi** [reply]
![smiling-face-with-hearts_1f970](https://github.com/ManimCommunity/manim/assets/8582807/38f7a254-3735-410b-80fd-28334bfe0fba)
